aw_transform/query2.py

```diff
@@ -0,0 +1,238 @@
+                raise QueryException("asd")
+            arg_end = arg_end - 1
+        # Parse name
+        name = string[:arg_start]
+        # Parse arguments
+        args = []
+        args_str = string[arg_start+1:arg_end-1]
+        while args_str:
+            arg, args_str = _parse_token(args_str, namespace)
+            comma = args_str.find(",")
+            if comma != -1:
+                args_str = args_str[comma+1:]
+            args.append(arg[0].parse(arg[1], namespace))
+        return Function(name, args)
+
+
+    def check(string: str):
+        i = 0
+        # Find opening bracket
+        found = False
+        for char in string:
+            if char.isalpha() or char == "_":
+                i = i+1
+            elif i != 0 and char.isdigit():
+                i = i+1
+            elif char == '(':
+                i = i+1
+                found = True
+                break
+            else:
+                break
+        if not found:
+            return None, string
+        to_consume = 1
+        for char in string[i:]:
+            i += 1
+            if char == ')':
+                to_consume = to_consume - 1
+            elif char == '(':
+                to_consume = to_consume + 1
+            if to_consume == 0:
+                break
+        if to_consume != 0:
+            raise QueryException("Unclosed function")
+        return string[:i], string[i+1:]
+
+def _parse_token(string: str, namespace: dict):
+    # TODO: The whole parsing thing is shoddily written, needs a rewrite from ground-up
+    if type(string) != str:
+        raise QueryException("Reached unreachable, cannot parse something that isn't a string")
+    if len(string) == 0:
+        return None
+    string = string.strip()
+    types = [String, Integer, Function, Variable]
+    token = None
+    t = None # Declare so we can return it
+    for t in types:
+        token, string = t.check(string)
+        if token:
+            break
+    if not token:
+        raise QueryException("Syntax error: {}".format(string))
+    print(t)
+    return (t, token), string
+
+def create_namespace() -> dict:
+    namespace = {
+        "NAME": None,
+        "CACHE": False,
+        "RETURN": None,
+        "TRUE": 1,
+        "FALSE": 0,
+        # start_time
+        # end_time
+    }
+    return namespace
+
+def parse(line, namespace):
+    separator_i = line.find("=")
+    var = line[:separator_i]
+    val = line[separator_i+1:]
+    if not val:
+        # TODO: Proper message
+        raise QueryException("Nothing to assign")
+    var, rest = _parse_token(var, namespace)
+    if var[0] is not Variable:
+        # TODO: Proper message
+        raise QueryException("Cannot assign to a non-variable")
+    val, rest = _parse_token(val, namespace)
+    # Parse token
+    var = var[0].parse(var[1], namespace)
+    val = val[0].parse(val[1], namespace)
+    return val, var
+
+def interpret(var, val, namespace, datastore):
+    if type(var) is Variable:
```

Comment:
Similar to the typecheck before but this works. Still better to use `isinstance` though.